x-pack/apm-server/sampling/pubsub/pubsub.go

```diff
@@ -0,0 +1,219 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package pubsub
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"github.com/pkg/errors"
+	"go.elastic.co/fastjson"
+
+	"github.com/elastic/beats/v7/libbeat/logp"
+	"github.com/elastic/go-elasticsearch/v7/esapi"
+	"github.com/elastic/go-elasticsearch/v7/esutil"
+
+	logs "github.com/elastic/apm-server/log"
+)
+
+// Pubsub provides a means of publishing and subscribing to sampled trace IDs,
+// using Elasticsearch for temporary storage.
+//
+// An independent process will periodically reap old documents in the index.
+type Pubsub struct {
+	config  Config
+	indexer esutil.BulkIndexer
+}
+
+// New returns a new Pubsub which can publish and subscribe sampled trace IDs,
+// using Elasticsearch for storage.
+//
+// Documents are expected to be indexed through a pipeline which sets the
+// `event.ingested` timestamp field. Another process will periodically reap
+// events older than a configured age.
+func New(config Config) (*Pubsub, error) {
+	if err := config.Validate(); err != nil {
+		return nil, errors.Wrap(err, "invalid pubsub config")
+	}
+	if config.Logger == nil {
+		config.Logger = logp.NewLogger(logs.Sampling)
+	}
+	indexer, err := esutil.NewBulkIndexer(esutil.BulkIndexerConfig{
+		Client:        config.Client,
+		Index:         config.Index,
+		FlushInterval: config.FlushInterval,
+		OnError: func(ctx context.Context, err error) {
+			config.Logger.With(logp.Error(err)).Debug("publishing sampled trace IDs failed")
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &Pubsub{config: config, indexer: indexer}, nil
+}
+
+// PublishSampledTraceIDs bulk indexes traceIDs into Elasticsearch.
+func (p *Pubsub) PublishSampledTraceIDs(ctx context.Context, traceID ...string) error {
+	for _, id := range traceID {
+		var doc traceIDDocument
+		doc.Observer.ID = p.config.BeatID
+		doc.Trace.ID = id
+
+		var json fastjson.Writer
+		if err := doc.MarshalFastJSON(&json); err != nil {
+			return err
+		}
+		if err := p.indexer.Add(ctx, esutil.BulkIndexerItem{
+			Action: "index",
+			Body:   bytes.NewReader(json.Bytes()),
```

Comment:
I was reluctant to add that because doing so leads to go-elasticsearch copying the body: https://github.com/elastic/go-elasticsearch/blob/e21ad8e37e8d947e01c8b0c426663c6120f9651e/esutil/bulk_indexer.go#L419-L427

I'll add it for now and reassess later.

Reply:
ah, I didn't realize that actually